src/main/java/io/confluent/avro/random/generator/Generator.java

```diff
@@ -1394,33 +1431,38 @@ private static long modulo(long first, long second) {
       FLOAT, DOUBLE
     }
 
-    private final double start;
-    private final double restart;
-    private final double step;
+    private final BigDecimal start;
+    private final BigDecimal restart;
+    private final BigDecimal modulo;
+    private final BigDecimal step;
     private final Type type;
-    private double current;
+    private BigDecimal current;
 
-    public DecimalIterator(double start, double restart, double step, Type type) {
-      this.start = start;
-      this.restart = restart;
-      this.step = step;
+    public DecimalIterator(double start, double restart, double step, long count, Type type) {
+      this.start = BigDecimal.valueOf(start);
+      this.restart = BigDecimal.valueOf(restart);
+      this.modulo = this.restart.subtract(this.start);
+      this.step = BigDecimal.valueOf(step);
       this.type = type;
-      current = start;
+      current = BigDecimal.ZERO;
+      if (count > 0) {
+        current = BigDecimal.valueOf(count)
+            .multiply(this.step)
+            .divideAndRemainder(this.modulo)[1];
+      }
     }
 
     @Override
     public Object next() {
-      double result = current;
-      if ((step > 0 && current >= restart - step) || (step < 0 && current <= restart - step)) {
-        current = start + modulo(step - (restart - current), restart - start);
-      } else {
-        current += step;
-      }
+      BigDecimal result = current.add(start);
+      current = current
+          .add(step)
+          .divideAndRemainder(modulo)[1];
```

Comment:
Can this be replaced with:
```suggestion
          .remainder(modulo);
```
?